packages/e2e/test/wallet/SingleAddressWallet/nft.test.ts

```diff
@@ -1,83 +1,163 @@
-        const keyAgent = await createKeyAgent(dependencies);
-        keyAgent.knownAddresses$ = of([
-          {
-            accountIndex: 0,
-            address: Cardano.Address(
-              // eslint-disable-next-line max-len
-              'addr_test1qpapna8hhj2hx2q2s9mdp7995a3hgundyvvvj0v0yq68lt8e6dgndgyep9stycsnejcnu8vm7a6dtqqhmf362z7fy5ksg46rum'
-            ),
-            index: 0,
-            networkId: Cardano.NetworkId.testnet,
-            rewardAccount: Cardano.RewardAccount('stake_test1uruax5fk5zvsjc9jvgfuevf7rkdlwax4sqta5ca9p0yj2tg4cf29e'),
-            type: KeyManagement.AddressType.External
-          }
-        ]);
-        return keyAgent;
+    wallet = (await getWallet({ env, idx: 0, logger, name: 'Minting Wallet', polling: { interval: 50 } })).wallet;
+
+    await firstValueFrom(wallet.syncStatus.isSettled$.pipe(filter((isSettled) => isSettled)));
+
+    const params = await firstValueFrom(wallet.genesisParameters$);
+
+    const keyAgent = await KeyManagement.InMemoryKeyAgent.fromBip39MnemonicWords(
+      {
+        getPassword: async () => Buffer.from(''),
+        mnemonicWords: KeyManagement.util.generateMnemonicWords(),
+        networkId: params.networkId
       },
-      createWallet: async (keyAgent) =>
-        new SingleAddressWallet(
-          {
-            name: 'Test Wallet'
-          },
-          {
-            assetProvider: await assetProviderFactory.create(env.ASSET_PROVIDER, env.ASSET_PROVIDER_PARAMS, logger),
-            chainHistoryProvider: await chainHistoryProviderFactory.create(
-              env.CHAIN_HISTORY_PROVIDER,
-              env.CHAIN_HISTORY_PROVIDER_PARAMS,
-              logger
-            ),
-            keyAgent,
-            logger,
-            networkInfoProvider: await networkInfoProviderFactory.create(
-              env.NETWORK_INFO_PROVIDER,
-              env.NETWORK_INFO_PROVIDER_PARAMS,
-              logger
-            ),
-            rewardsProvider: await rewardsProviderFactory.create(
-              env.REWARDS_PROVIDER,
-              env.REWARDS_PROVIDER_PARAMS,
-              logger
-            ),
-            stakePoolProvider: await stakePoolProviderFactory.create(
-              env.STAKE_POOL_PROVIDER,
-              env.STAKE_POOL_PROVIDER_PARAMS,
-              logger
-            ),
-            txSubmitProvider: await txSubmitProviderFactory.create(
-              env.TX_SUBMIT_PROVIDER,
-              env.TX_SUBMIT_PROVIDER_PARAMS,
-              logger
-            ),
-            utxoProvider: await utxoProviderFactory.create(env.UTXO_PROVIDER, env.UTXO_PROVIDER_PARAMS, logger)
+      { inputResolver: { resolveInputAddress: async () => null } }
+    );
+
+    const derivedAddress = await keyAgent.deriveAddress({
+      index: 0,
+      type: KeyManagement.AddressType.External
+    });
+
+    const keyHash = Cardano.policyKeyHash(derivedAddress.address);
```

Comment:
We cant derive the key hash just from the public key as it doesnt have enough information; the key hash needs the stake credential, the network id, and the pub key, which are all encoded in the address. If you notice the RewardAccount which is used in the other function is also an address (the staking address). I will, however, move that utility function to the place you suggest.

Reply:
From `Cardano.policyKeyHash`:

> BaseAddress.from_address(cslAddress)!.payment_cred().to_keyhash()

Do we really need stake credential and network ID here? `payment_cred().to_keyhash()` should equal to just payment key hash as far as I understand

Reply:
I'm wondering why `payment` is of type `StakeCredential` :thinking: is this from CSL?

Anyway, my point was that you don't need to create a grouped address just to get the key hash of payment key. `derivePublicKey` should be sufficient. Or maybe I'm missing something here, let's chat about this tomorrow.

Reply:
Yes that's from the CSL; you are right, otherwise, the key hash would change between networks and it doesn't; I will dig into the CSL code and see how the payment_cred is actually being calculated.

Reply:
addressed here: https://github.com/input-output-hk/cardano-js-sdk/pull/404/commits/96640f3b9db0271f290550eeb86a2f9809ee9358